File: src/FrostAura.Libraries.Components/Presentational/Map/GoogleMap.razor.js
window.faGoogleMap = {
    // Maps collection key/value pair.
    maps: {},
    // Markers collection key (map id) / value pair (collection of markers).
    markers: {},
    // Initialize a Google Map and set its zoom and center.
    initializeGoogleMapAsync: async (request) => {
        await window
            .faGoogleMap
            .boostrapMapAsync(request);
    },
    // Wire up the map in it's entirety async.
    boostrapMapAsync: async (request) => {
        await window.faGoogleMap.loadSdkAsync(request.apiKey);
        await window.faGoogleMap.initializeMapAsync(request);
    },
    // Load the Google Maps SDK script if it hasn't already been loaded.
    loadSdkAsync: (apiKey) => {
        const scriptId = 'googleMapsScript';
        const response = new Promise((resolve, reject) => {
            if (!!document.body.querySelector(`#${scriptId}`)) {
                resolve();

                return;
            }

            if (!apiKey) {
                reject('No api key was specified for component fa-google-map.');

                return;
            }

            const ele = document
                .createElement('script');

            ele.id = scriptId;
            ele.src = `https://maps.googleapis.com/maps/api/js?key=${apiKey}`;
            ele.onload = () => resolve();
            ele.onerror = () => reject();

            document.body.appendChild(ele);
        });

        return response;
    },
    // Initialize the map drivers.
    initializeMapAsync: async (request) => {
        const mapElement = document.querySelector('fa-google-map > div[id="' + request.id + '"]');

        if (!mapElement) throw new Error('No map element found.');

        await Promise.all([window.faGoogleMap.waitForComponentStylingAsync(request), window.faGoogleMap.waitForGoogleSdkAsync()]);

        const parentComponent = mapElement.parentElement;
        const width = parentComponent.offsetWidth;
        const height = parentComponent.offsetHeight;

        mapElement.style.width = `${width}px`;
        mapElement.style.height = `${height}px`;

        window.faGoogleMap.maps[request.id] = new google.maps.Map(mapElement, {
            center: { lat: parseFloat(request.center.lat), lng: parseFloat(request.center.lng) },
            zoom: request.zoom,
            mapTypeId: request.mapType,
            gestureHandling: 'cooperative',
            disableDefaultUI: true
        });
    },
    // Wait for the component's styling to kick in async.
    waitForComponentStylingAsync: async (request) => {
        const checkWidthRecursively = (resolve) => {
            const mapElement = document.querySelector('fa-google-map > div[id="' + request.id + '"]');
            const parentComponent = mapElement.parentElement;
            const width = parentComponent.offsetWidth;

            if (!width || width <= 0) {
                setTimeout(() => checkWidthRecursively(resolve));

                return;
            }

            resolve();
        };
        const response = new Promise((resolve) => {
            checkWidthRecursively(resolve);
        });

        return response;
    },
    // Wait for the Google SDK to kick in async.
    waitForGoogleSdkAsync: async () => {
        const ensureGoogleSdk = (resolve) => {
            const isGoogleSdkAvailable = !!window.google && !!window.google.maps && !!window.google.maps.Map;

            if (isGoogleSdkAvailable) {
                resolve();

                return;
            }

            setTimeout(() => ensureGoogleSdk(resolve));
        };
        const response = new Promise((resolve) => {
            ensureGoogleSdk(resolve);
        });

        return response;
    },
    // Add a marker to the map.
    addMarker: (mapId, request) => {
        window.faGoogleMap.markers[mapId] = window.faGoogleMap.markers[mapId] || [];

        const map = window.faGoogleMap.maps[mapId];
        const marker = new google.maps.Marker({
            position: request.location,
            title: request.title,
            map,
            icon: {
                url: request.icon.url,
                scaledSize: new google.maps.Size(request.icon.size, request.icon.size)
            }
        });
        const infoWindow = new google.maps.InfoWindow({});

        window.faGoogleMap.markers[mapId].push(marker);
        google.maps.event.addListener(marker, 'click', function () {
            infoWindow.setContent(request.info.content);
            infoWindow.open(window.faGoogleMap.maps[mapId], marker);
        });
    },
    // Clear all markers from the map.
    clearAllMarkers: (mapId) => {
        const markers = window.faGoogleMap.markers[mapId] = window.faGoogleMap.markers[mapId] || [];

        for (let i = 0; i < markers.length; i++) {
            markers[i].setMap(null);
        }

        window.faGoogleMap.markers[mapId] = [];
    }
}

File: src/FrostAura.Libraries.Components/Presentational/Map/GoogleMap.razor.cs
using FrostAura.Libraries.Components.Abstractions;
using FrostAura.Libraries.Components.Shared.Enums.Map;
using FrostAura.Libraries.Components.Shared.Models.Map;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace FrostAura.Libraries.Components.Presentational.Map
{
    /// <summary>
    /// Google map component.
    /// </summary>
    public partial class GoogleMap : BaseComponent<GoogleMap>
    {
        /// <summary>
        /// The current version of the component.
        /// </summary>
        public override Version Version { get; } = new Version(0, 0, 1);
        /// <summary>
        /// Center point for the map.
        /// </summary>
        [Parameter]
        public GeoPoint Center { get; set; }
        /// <summary>
        /// Default zoom of the map.
        /// </summary>
        [Parameter]
        public int Zoom { get; set; }
        /// <summary>
        /// Map type to render initially.
        /// </summary>
        [Parameter]
        public MapType MapType { get; set; }
        /// <summary>
        /// API key to use with the map service.
        /// </summary>
        [Parameter]
        public string ApiKey { get; set; }
        /// <summary>
        /// Event source for when the map is initialized.
        /// </summary>
        [Parameter]
        public EventCallback<GoogleMap> OnMapReady { get; set; }

        /// <summary>
        /// Add a marker to the map.
        /// </summary>
        public async Task AddMarkerAsync(string title,
            string subtitle,
            double lat,
            double lng,
            int size = 25,
            string iconUrl = "https://wiredscore.com/wp-content/uploads/2017/02/pin-drop-transparent-white.png")
        {
            var request = new
            {
                title,
                icon = new
                {
                    url = iconUrl,
                    size
                },
                location = new
                {
                    lat,
                    lng
                },
                info = new
                {
                    content = @"
                        <div>
                            <div style='font-weight: bold'>" + title + @"</div>
                            <div>" + subtitle + @"</div>
                        </div>"
                }
            };

            await JsRuntime.InvokeVoidAsync("faGoogleMap.addMarker", Id, request);
        }

        /// <summary>
        /// Clear all markers from the map.
        /// </summary>
        public async Task ClearAllMarkersAsync()
        {
            await JsRuntime.InvokeVoidAsync("faGoogleMap.clearAllMarkers", Id);
        }

        /// <summary>
        /// Initialize the Google map component.
        /// </summary>
        private async Task InitializeMapAsync()
        {
            await JsRuntime.InvokeVoidAsync("faGoogleMap.initializeGoogleMapAsync", new
            {
                id = Id,
                center = new
                {
                    lat = Center?.Latitude,
                    lng = Center?.Longitude
                },
                zoom = Zoom,
                mapType = MapType.ToString().ToLower(),
                apiKey = ApiKey
            });
            await OnMapReady.InvokeAsync(this);
        }
    }
}
